handlers/handler-router/src/main/java/org/apache/servicecomb/router/cache/RouterRuleCache.java

```diff
@@ -0,0 +1,126 @@
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.servicecomb.router.model.PolicyRuleItem;
+import org.apache.servicecomb.router.model.ServiceInfoCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+import org.yaml.snakeyaml.Yaml;
+
+/**
+ * @Author GuoYl123
+ * @Date 2019/10/17
+ **/
+public class RouterRuleCache {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RouterRuleCache.class);
+
+  private static ConcurrentHashMap<String, ServiceInfoCache> serviceInfoCacheMap = new ConcurrentHashMap<>();
+
+  private static final String ROUTE_RULE = "servicecomb.routeRule.%s";
+
+  private static Interner<String> servicePool = Interners.newWeakInterner();
+
+  /**
+   * cache and register callback return false when: 1. parsing error 2. rule is null
+   *
+   * @param targetServiceName
+   * @return
+   */
+  public static boolean doInit(String targetServiceName) {
+    if (!isServerContainRule(targetServiceName)) {
+      return false;
+    }
+    if (!serviceInfoCacheMap.containsKey(targetServiceName)) {
+      synchronized (servicePool.intern(targetServiceName)) {
+        if (serviceInfoCacheMap.containsKey(targetServiceName)) {
+          return true;
+        }
+        //Yaml not thread-safe
+        DynamicStringProperty ruleStr = DynamicPropertyFactory.getInstance().getStringProperty(
+            String.format(ROUTE_RULE, targetServiceName), null, () -> {
+              refresh(targetServiceName);
+              DynamicStringProperty tepRuleStr = DynamicPropertyFactory.getInstance()
+                  .getStringProperty(String.format(ROUTE_RULE, targetServiceName), null);
+              addAllRule(targetServiceName, tepRuleStr.get());
+            });
+        return addAllRule(targetServiceName, ruleStr.get());
+      }
+    }
+    return true;
+  }
+
+  private static boolean addAllRule(String targetServiceName, String ruleStr) {
+    if (StringUtils.isEmpty(ruleStr)) {
+      return false;
+    }
+    Yaml yaml = new Yaml();
+    List<PolicyRuleItem> policyRuleItemList;
+    try {
+      policyRuleItemList = Arrays
+          .asList(yaml.loadAs(ruleStr, PolicyRuleItem[].class));
+    } catch (Exception e) {
+      LOGGER.error("route management Serialization failed: {}", e.getMessage());
+      return false;
+    }
+    if (CollectionUtils.isEmpty(policyRuleItemList)) {
+      return false;
+    }
+    ServiceInfoCache serviceInfoCache = new ServiceInfoCache(policyRuleItemList);
+    serviceInfoCacheMap.put(targetServiceName, serviceInfoCache);
+    return true;
+  }
+
+  /**
+   * if a server don't have rule , avoid registered too many callback , it may cause memory leak
+   *
+   * @param targetServiceName
+   * @return
+   */
+  public static boolean isServerContainRule(String targetServiceName) {
+    DynamicStringProperty lookFor = DynamicPropertyFactory.getInstance()
```

Comment:
targetServiceName is checked before going to loadbalancer, so it will NOT causs OOM normally. If do not watch, how about when create a rule and then delete the rule?  Will this condition cause problem?